### ISSUE TITLE
Syntax: [Symbol Table] Maintain tables by variable data type

### ISSUE THREAD

**Issue by meganindya** [sugarlabs/musicblocks-v4-lib#84]
## Description

The **Symbol Table** component in `src/syntax/symbol-table/symbolTable.ts` maintains *global*, *program-specific*, and *routine-specific* variables only by *variable name*. You can only fetch variables by *variable name*. Add methods for each table to fetch variables by *variable name* and *variable data type*.

For example, say there is a global variable `foo` of data type `number`. Currently you can fetch it by `getGlobalVariable('foo')`. The objective is to optionally supply the type as well like `getGlobalVariable('foo', 'number')`, which returns the same. But, calling `getGlobalVariable('foo', 'string')` returns `null`.

## Objectives

Refactor function signatures (make change in `symbolTable.d.ts` file and the `symbolTable.ts` file):

- [ ] `getGlobalVariable(variable: string, dataType?: TDataName): IVariable | null`
- [ ] `getProgramVariable(variable: string, program: string, dataType?: TDataName): IVariable | null`
- [ ] `getRoutineVariable(variable: string, routine: string, dataType?: TDataName): IVariable | null`

Add unit tests for:

- [ ] `getGlobalVariable` with `dataType` field
- [ ] `getProgramVariable` with `dataType` field
- [ ] `getRoutineVariable` with `dataType` field

**Comment by S-kus** [comment]
Hello @meganindya, I want to work on this issue.
After going through the code, the file `symbolTable.d.ts` and `symbolTable.ts` are currently in `src/execution/`, Ig it's needed to update in the description of the issue.
Could you please confirm once again that we needed to change the `getGlobalVariable()` function similar to `addGlobalVariable()` and what is `getProgramVariable` I am not able to find it in any files, Is it `getProcessVariable`? but it's already updated according to the issue.

**Comment by meganindya** [comment]
Yes, I updated the identifiers. The output variable format needs to be what the `IVariable` type describes. However, all three private variable tables `_globalTable`, `_processTable`, and `_routineTable` store variables by variable name. They need to store those variables by data types at the top level and then by variable names inside the data types namespace. This way two identically named variables with different data types can be stored.

The objective is to:
- make changes in the private variable tables to store variables by data type and then by variable names within them
- make changes to the public query functions as specified above

**Comment by S-kus** [comment]
Okay @meganindya, working on it.

**Comment by impAkshat** [comment]
Do I need to refactor other functions like `_addTableVariable`, `_getTableVariable` also?
